supervisor panics if ide is restarted

```
{"message":"IDE is ready","serviceContext":{"service":"supervisor","version":""},"severity":"info","time":"2020-10-30T13:01:54Z"}
panic: close of closed channel

goroutine 218 [running]:
github.com/gitpod-io/gitpod/supervisor/pkg/supervisor.runIDEReadinessProbe(0xc0003cd440, 0xc0003d3320)
	/tmp/build/components-supervisor--app.402afef9f129581793a26b7b492cc2c74a4d3db2/pkg/supervisor/supervisor.go:430 +0x645
created by github.com/gitpod-io/gitpod/supervisor/pkg/supervisor.startAndWatchIDE
	/tmp/build/components-supervisor--app.402afef9f129581793a26b7b492cc2c74a4d3db2/pkg/supervisor/supervisor.go:306 +0x5ea
```

It leads to the stopped workspace. cc @csweichel

--- a/components/ide/code/leeway.Dockerfile
+++ b/components/ide/code/leeway.Dockerfile
@@ -28,7 +28,7 @@ RUN sudo apt-get update \
     && sudo apt-get clean -y \
     && rm -rf /var/lib/apt/lists/*
 
-ENV GP_CODE_COMMIT 02021b3f4267917afb774747dcca21601fce52cb
+ENV GP_CODE_COMMIT 8a139e627387c9abd52d1161288fe019903e3270
 RUN mkdir gp-code \
     && cd gp-code \
     && git init \

--- a/components/supervisor/pkg/supervisor/services.go
+++ b/components/supervisor/pkg/supervisor/services.go
@@ -38,11 +38,49 @@ type RegisterableRESTService interface {
 	RegisterREST(mux *runtime.ServeMux, grpcEndpoint string) error
 }
 
+type ideReadyState struct {
+	ready bool
+	cond  *sync.Cond
+}
+
+// Wait returns a channel that emits when IDE is ready
+func (service *ideReadyState) Wait() <-chan struct{} {
+	ready := make(chan struct{})
+	go func() {
+		service.cond.L.Lock()
+		for !service.ready {
+			service.cond.Wait()
+		}
+		service.cond.L.Unlock()
+		close(ready)
+	}()
+	return ready
+}
+
+// Get checks whether IDE is ready
+func (service *ideReadyState) Get() bool {
+	service.cond.L.Lock()
+	ready := service.ready
+	service.cond.L.Unlock()
+	return ready
+}
+
+// Set updates IDE ready state
+func (service *ideReadyState) Set(ready bool) {
+	service.cond.L.Lock()
+	defer service.cond.L.Unlock()
+	if service.ready == ready {
+		return
+	}
+	service.ready = ready
+	service.cond.Broadcast()
+}
+
 type statusService struct {
 	ContentState ContentState
 	Ports        *ports.Manager
 	Tasks        *tasksManager
-	IDEReady     <-chan struct{}
+	ideReady     *ideReadyState
 }
 
 func (s *statusService) RegisterGRPC(srv *grpc.Server) {
@@ -60,20 +98,14 @@ func (s *statusService) SupervisorStatus(context.Context, *api.SupervisorStatusR
 func (s *statusService) IDEStatus(ctx context.Context, req *api.IDEStatusRequest) (*api.IDEStatusResponse, error) {
 	if req.Wait {
 		select {
-		case <-s.IDEReady:
+		case <-s.ideReady.Wait():
 			return &api.IDEStatusResponse{Ok: true}, nil
 		case <-ctx.Done():
 			return nil, status.Error(codes.DeadlineExceeded, ctx.Err().Error())
 		}
 	}
 
-	var ok bool
-	select {
-	case <-s.IDEReady:
-		ok = true
-	default:
-		ok = false
-	}
+	ok := s.ideReady.Get()
 	return &api.IDEStatusResponse{Ok: ok}, nil
 }
 

--- a/components/supervisor/pkg/supervisor/supervisor.go
+++ b/components/supervisor/pkg/supervisor/supervisor.go
@@ -114,7 +114,7 @@ func Run(options ...RunOption) {
 	ctx, cancel := context.WithCancel(context.Background())
 	var (
 		shutdown            = make(chan struct{})
-		ideReady            = make(chan struct{})
+		ideReady            = &ideReadyState{cond: sync.NewCond(&sync.Mutex{})}
 		cstate              = NewInMemoryContentState(cfg.RepoRoot)
 		gitpodService       = createGitpodService(cfg, tokenService)
 		gitpodConfigService = gitpod.NewConfigService(cfg.RepoRoot+"/.gitpod.yml", cstate.ContentReady())
@@ -139,7 +139,7 @@ func Run(options ...RunOption) {
 			ContentState: cstate,
 			Ports:        portMgmt,
 			Tasks:        taskManager,
-			IDEReady:     ideReady,
+			ideReady:     ideReady,
 		},
 		termMuxSrv,
 		RegistrableTokenService{tokenService},
@@ -272,7 +272,7 @@ func hasMetadataAccess() bool {
 	return false
 }
 
-func startAndWatchIDE(ctx context.Context, cfg *Config, wg *sync.WaitGroup, ideReady chan<- struct{}) {
+func startAndWatchIDE(ctx context.Context, cfg *Config, wg *sync.WaitGroup, ideReady *ideReadyState) {
 	defer wg.Done()
 
 	type status int
@@ -303,14 +303,17 @@ supervisorLoop:
 			}
 			s = statusShouldRun
 
-			go runIDEReadinessProbe(cfg, ideReady)
+			go func() {
+				runIDEReadinessProbe(cfg)
+				ideReady.Set(true)
+			}()
 
 			go func() {
 				err := cmd.Wait()
 				if err != nil && !strings.Contains(err.Error(), "signal: interrupt") {
 					log.WithError(err).Warn("IDE was stopped")
 				}
-
+				ideReady.Set(false)
 				close(ideStopped)
 			}()
 		}
@@ -399,8 +402,7 @@ func buildIDEEnv(cfg *Config) []string {
 	return env
 }
 
-func runIDEReadinessProbe(cfg *Config, ideReady chan<- struct{}) {
-	defer close(ideReady)
+func runIDEReadinessProbe(cfg *Config) {
 	defer log.Info("IDE is ready")
 
 	switch cfg.ReadinessProbe.Type {

--- a/components/theia/packages/gitpod-extension/src/browser/gitpod-task-contribution.ts
+++ b/components/theia/packages/gitpod-extension/src/browser/gitpod-task-contribution.ts
@@ -4,19 +4,18 @@
  * See License-AGPL.txt in the project root for license information.
  */
 
-import { injectable, inject, postConstruct } from 'inversify';
 import { FrontendApplicationContribution } from '@theia/core/lib/browser';
-import { TerminalFrontendContribution } from '@theia/terminal/lib/browser/terminal-frontend-contribution';
+import { Deferred } from '@theia/core/lib/common/promise-util';
 import { TerminalWidget } from '@theia/terminal/lib/browser/base/terminal-widget';
-import { GitpodTerminalWidget } from './gitpod-terminal-widget';
-import { GitpodTaskState, GitpodTaskServer, GitpodTask } from '../common/gitpod-task-protocol';
+import { TerminalFrontendContribution } from '@theia/terminal/lib/browser/terminal-frontend-contribution';
 import { IBaseTerminalServer } from '@theia/terminal/lib/common/base-terminal-protocol';
-import { Emitter } from '@theia/core';
+import { inject, injectable, postConstruct } from 'inversify';
+import { GitpodTask, GitpodTaskServer, GitpodTaskState } from '../common/gitpod-task-protocol';
+import { GitpodTerminalWidget } from './gitpod-terminal-widget';
 
 interface GitpodTaskTerminalWidget extends GitpodTerminalWidget {
     readonly kind: 'gitpod-task'
-    /** undefined if not running */
-    remoteTerminal?: string
+    applyTask(updated: GitpodTask): Promise<void>
 }
 namespace GitpodTaskTerminalWidget {
     const idPrefix = 'gitpod-task-terminal'
@@ -40,16 +39,20 @@ export class GitpodTaskContribution implements FrontendApplicationContribution {
     @inject(GitpodTaskServer)
     private readonly server: GitpodTaskServer;
 
-    private readonly onDidChangeEmitter = new Emitter<GitpodTask[]>();
-    private readonly onDidChange = this.onDidChangeEmitter.event;
-
     private readonly taskTerminals = new Map<string, GitpodTaskTerminalWidget>();
+
+    private readonly pendingInitialTasks = new Deferred<GitpodTask[]>();
+    private readonly pendingInitializeLayout = new Deferred<void>();
+    private updateQueue = this.pendingInitializeLayout.promise;
 
     @postConstruct()
     protected init(): void {
         // register client before connection is opened
         this.server.setClient({
-            onDidChange: ({ updated }) => this.onDidChangeEmitter.fire(updated)
+            onDidChange: ({ updated }) => {
+                this.pendingInitialTasks.resolve(updated);
+                this.queue(() => this.updateTerminals(updated));
+            }
         });
         this.terminals.onDidCreateTerminal(terminal => {
             if (GitpodTaskTerminalWidget.is(terminal)) {
@@ -57,19 +60,60 @@ export class GitpodTaskContribution implements FrontendApplicationContribution {
                 terminal.onDidDispose(() =>
                     this.taskTerminals.delete(terminal.id)
                 );
+                const attachTerminal = terminal['attachTerminal'].bind(terminal);
+                terminal['attachTerminal'] = id => attachTerminal(id).then(terminalId => {
+                    if (IBaseTerminalServer.validateId(terminalId)) {
+                        return terminalId
+                    }
+                    return terminal['createTerminal']()
+                })
+                let starting = false;
+                const start = terminal['start'].bind(terminal);
+                terminal['start'] = async id => {
+                    starting = true;
+                    try {
+                        return start(id)
+                    } finally {
+                        starting = false;
+                    }
+                }
+
+                let task: GitpodTask | undefined;
+                const update = async () => {
+                    if (task?.state === GitpodTaskState.CLOSED) {
+                        terminal.dispose();
+                        return;
+                    }
+                    const remoteTerminal = task?.terminal;
+                    if (task?.state !== GitpodTaskState.RUNNING || !remoteTerminal || starting) {
+                        return;
+                    }
+                    const terminalId = terminal.terminalId;
+                    if (IBaseTerminalServer.validateId(terminalId)) {
+                        await this.server.attach({ terminalId, remoteTerminal });
+                    } else {
+                        await terminal.start();
+                    }
+                }
+                terminal.applyTask = updated => {
+                    task = updated
+                    return update()
+                }
                 terminal.onTerminalDidClose(() => {
-                    if (terminal.remoteTerminal) {
-                        fetch(window.location.protocol + '//' + window.location.host + '/_supervisor/v1/terminal/close/' + terminal.remoteTerminal, {
+                    if (task && task.terminal && task.state !== GitpodTaskState.CLOSED) {
+                        fetch(window.location.protocol + '//' + window.location.host + '/_supervisor/v1/terminal/close/' + task.terminal, {
                             credentials: "include"
                         })
                     }
                 });
+                terminal.onDidOpen(update)
+                terminal.onDidOpenFailure(update);
             }
         });
     }
 
     async onDidInitializeLayout(): Promise<void> {
-        const tasks = await this.server.getTasks();
+        const tasks = await this.pendingInitialTasks.promise;
         let ref: TerminalWidget | undefined;
         for (const task of tasks) {
             if (task.state == GitpodTaskState.CLOSED) {
@@ -79,30 +123,24 @@ export class GitpodTaskContribution implements FrontendApplicationContribution {
                 const id = GitpodTaskTerminalWidget.toTerminalId(task.id);
                 let terminal = this.taskTerminals.get(id);
                 if (!terminal) {
-                    terminal = await this.terminals.newTerminal({
+                    terminal = (await this.terminals.newTerminal({
                         id,
                         kind: 'gitpod-task',
                         title: task.presentation!.name,
                         useServerTitle: false
-                    }) as GitpodTaskTerminalWidget;
-                    await terminal.start();
+                    })) as GitpodTaskTerminalWidget;
                     this.terminals.activateTerminal(terminal, {
                         ref,
                         area: task.presentation.openIn || 'bottom',
                         mode: task.presentation.openMode || 'tab-after'
                     });
-                } else if (!IBaseTerminalServer.validateId(terminal.terminalId)) {
-                    await terminal.start();
                 }
-                if (terminal) {
-                    ref = terminal;
-                }
+                ref = terminal;
             } catch (e) {
                 console.error('Failed to start Gitpod task terminal:', e);
             }
         }
-        this.updateTerminals(tasks);
-        this.onDidChange(tasks => this.updateTerminals(tasks));
+        this.pendingInitializeLayout.resolve();
 
         // if there is no terminal at all, lets start one
         if (!this.terminals.all.length) {
@@ -112,7 +150,7 @@ export class GitpodTaskContribution implements FrontendApplicationContribution {
         }
     }
 
-    protected async updateTerminals(tasks: GitpodTask[]): Promise<void> {
+    private async updateTerminals(tasks: GitpodTask[]): Promise<void> {
         for (const task of tasks) {
             try {
                 const id = GitpodTaskTerminalWidget.toTerminalId(task.id);
@@ -120,24 +158,15 @@ export class GitpodTaskContribution implements FrontendApplicationContribution {
                 if (!terminal) {
                     continue;
                 }
-                if (task.state === GitpodTaskState.CLOSED) {
-                    delete terminal.remoteTerminal;
-                    terminal.dispose();
-                    continue;
-                }
-                if (task.state !== GitpodTaskState.RUNNING) {
-                    continue;
-                }
-                terminal.remoteTerminal = task.terminal;
-                if (task.terminal) {
-                    await this.server.attach({
-                        terminalId: terminal.terminalId,
-                        remoteTerminal: task.terminal
-                    });
-                }
+                await terminal.applyTask(task);
             } catch (e) {
                 console.error('Failed to update Gitpod task terminal:', e);
             }
         }
     }
+
+    private queue(update: () => Promise<void>): Promise<void> {
+        return this.updateQueue = this.updateQueue.then(update, update);
+    }
+
 }

--- a/components/theia/packages/gitpod-extension/src/browser/ports/gitpod-ports-service.ts
+++ b/components/theia/packages/gitpod-extension/src/browser/ports/gitpod-ports-service.ts
@@ -47,19 +47,20 @@ export class GitpodPortsService {
         this.server.setClient({
             onDidChange: event => this.updatePorts(event)
         });
-        this.server.getPorts().then(added => this.updatePorts({ added }));
     }
 
     get ports(): IterableIterator<PortsStatus.AsObject> {
         return this._ports.values();
     }
 
-    private updatePorts({ added, updated, removed }: DidChangeGitpodPortsEvent): void {
+    private updatePorts({ added, updated, removed, initial }: DidChangeGitpodPortsEvent): void {
+        const toClean = initial ? new Set<number>(this._ports.keys()) : undefined;
         for (const ports of [added, updated]) {
             if (ports === undefined) {
                 continue;
             }
             for (const port of ports) {
+                toClean?.delete(port.localPort)
                 const current = this._ports.get(port.localPort);
                 this._ports.set(port.localPort, port);
                 if (isExposedServedPort(port) && !isExposedServedPort(current)) {
@@ -67,8 +68,11 @@ export class GitpodPortsService {
                 }
             }
         }
-        if (removed) {
-            for (const port of removed) {
+        for (const ports of [removed, toClean]) {
+            if (ports === undefined) {
+                continue;
+            }
+            for (const port of ports) {
                 this._ports.delete(port);
             }
         }

--- a/components/theia/packages/gitpod-extension/src/common/gitpod-port-server.ts
+++ b/components/theia/packages/gitpod-extension/src/common/gitpod-port-server.ts
@@ -11,7 +11,6 @@ export const gitpodPortServicePath = '/services/gitpodPorts';
 
 export const GitpodPortServer = Symbol('GitpodPortServer');
 export interface GitpodPortServer extends JsonRpcServer<GitpodPortClient> {
-    getPorts(): Promise<PortsStatus.AsObject[]>;
     exposePort(params: ExposeGitpodPortParams): Promise<void>;
 }
 
@@ -20,6 +19,7 @@ export interface GitpodPortClient {
 }
 
 export interface DidChangeGitpodPortsEvent {
+    initial?: boolean
     added?: PortsStatus.AsObject[]
     updated?: PortsStatus.AsObject[]
     removed?: number[]

--- a/components/theia/packages/gitpod-extension/src/common/gitpod-task-protocol.ts
+++ b/components/theia/packages/gitpod-extension/src/common/gitpod-task-protocol.ts
@@ -28,7 +28,6 @@ export const gitpodTaskServicePath = "/services/gitpodTasks";
 
 export const GitpodTaskServer = Symbol('GitpodTaskServer');
 export interface GitpodTaskServer extends JsonRpcServer<GitpodTaskClient> {
-    getTasks(): Promise<GitpodTask[]>;
     attach(params: AttachTaskTerminalParams): Promise<void>
 }
 export interface GitpodTaskClient {

--- a/components/theia/packages/gitpod-extension/src/node/gitpod-backend-module.ts
+++ b/components/theia/packages/gitpod-extension/src/node/gitpod-backend-module.ts
@@ -60,7 +60,6 @@ export default new ContainerModule((bind, unbind, isBound, rebind) => {
         new JsonRpcConnectionHandler<GitpodPortClient>(gitpodPortServicePath, client => {
             const server = context.container.get<GitpodPortServerImpl>(GitpodPortServer);
             server.setClient(client);
-            client.onDidCloseConnection(() => server.disposeClient(client));
             return server;
         })
     ).inSingletonScope();
@@ -69,7 +68,6 @@ export default new ContainerModule((bind, unbind, isBound, rebind) => {
         new JsonRpcConnectionHandler<GitpodTaskClient>(gitpodTaskServicePath, client => {
             const server = context.container.get<GitpodTaskServerImpl>(GitpodTaskServer);
             server.setClient(client);
-            client.onDidCloseConnection(() => server.disposeClient(client));
             return server;
         })
     ).inSingletonScope();

--- a/components/theia/packages/gitpod-extension/src/node/gitpod-port-server-impl.ts
+++ b/components/theia/packages/gitpod-extension/src/node/gitpod-port-server-impl.ts
@@ -10,6 +10,8 @@ import { inject, injectable, postConstruct } from 'inversify';
 import { GitpodPortClient, GitpodPortServer, ExposeGitpodPortParams } from '../common/gitpod-port-server';
 import { SupervisorClientProvider } from './supervisor-client-provider';
 import { ExposePortRequest, ExposePortResponse } from '@gitpod/supervisor-api-grpc/lib/control_pb';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { JsonRpcProxy } from '@gitpod/gitpod-protocol/lib/messaging/proxy-factory';
 
 @injectable()
 export class GitpodPortServerImpl implements GitpodPortServer {
@@ -23,11 +25,13 @@ export class GitpodPortServerImpl implements GitpodPortServer {
     private stopUpdates: (() => void) | undefined;
 
     private readonly ports = new Map<number, PortsStatus.AsObject>();
+    private readonly deferredReady = new Deferred<void>();
 
     @postConstruct()
     async start(): Promise<void> {
         const client = await this.supervisorClientProvider.getStatusClient();
         while (this.run) {
+            let initial = true;
             try {
                 const req = new PortsStatusRequest();
                 req.setObserve(true);
@@ -38,6 +42,10 @@ export class GitpodPortServerImpl implements GitpodPortServer {
                     evts.on('close', resolve);
                     evts.on('error', reject);
                     evts.on('data', (update: PortsStatusResponse) => {
+                        if (initial) {
+                            this.ports.clear();
+                            initial = false;
+                        }
                         let added: PortsStatus.AsObject[] | undefined
                         for (const port of update.getAddedList()) {
                             const object = port.toObject();
@@ -58,6 +66,7 @@ export class GitpodPortServerImpl implements GitpodPortServer {
                         for (const client of this.clients) {
                             client.onDidChange({ added, updated, removed });
                         }
+                        this.deferredReady.resolve();
                     });
                 });
             } catch (err) {
@@ -65,10 +74,6 @@ export class GitpodPortServerImpl implements GitpodPortServer {
                 await new Promise(resolve => setTimeout(resolve, 1000));
             }
         }
-    }
-
-    async getPorts(): Promise<PortsStatus.AsObject[]> {
-        return [...this.ports.values()];
     }
 
     async exposePort(params: ExposeGitpodPortParams): Promise<void> {
@@ -81,11 +86,22 @@ export class GitpodPortServerImpl implements GitpodPortServer {
         await util.promisify<ExposePortRequest, ExposePortResponse>(controlClient.exposePort).bind(controlClient)(request);
     }
 
-    setClient(client: GitpodPortClient): void {
-        this.clients.add(client);
-    }
-    disposeClient(client: GitpodPortClient): void {
-        this.clients.delete(client);
+    setClient(client: JsonRpcProxy<GitpodPortClient>): void {
+        let closed = false;
+        this.deferredReady.promise.then(() => {
+            if (closed) {
+                return;
+            }
+            this.clients.add(client);
+            client.onDidChange({
+                initial: true,
+                added: [...this.ports.values()]
+            })
+        });
+        client.onDidCloseConnection(() => {
+            closed = true;
+            this.clients.delete(client);
+        });
     }
 
     dispose(): void {

--- a/components/theia/packages/gitpod-extension/src/node/gitpod-task-server-impl.ts
+++ b/components/theia/packages/gitpod-extension/src/node/gitpod-task-server-impl.ts
@@ -5,6 +5,7 @@
  */
 
 import { TasksStatusRequest, TasksStatusResponse } from '@gitpod/supervisor-api-grpc/lib/status_pb';
+import { JsonRpcProxy } from '@theia/core/lib/common/messaging/proxy-factory';
 import { ApplicationShell } from '@theia/core/lib/browser';
 import { Deferred } from '@theia/core/lib/common/promise-util';
 import { ProcessManager, TerminalProcess } from '@theia/process/lib/node';
@@ -79,10 +80,10 @@ export class GitpodTaskServerImpl implements GitpodTaskServer {
                             this.tasks.set(task.getId(), update);
                             updated.push(update);
                         }
-                        this.deferredReady.resolve();
                         for (const client of this.clients) {
                             client.onDidChange({ updated });
                         }
+                        this.deferredReady.resolve();
                     });
                 });
             } catch (err) {
@@ -90,11 +91,6 @@ export class GitpodTaskServerImpl implements GitpodTaskServer {
                 await new Promise(resolve => setTimeout(resolve, 1000));
             }
         }
-    }
-
-    async getTasks(): Promise<GitpodTask[]> {
-        await this.deferredReady.promise;
-        return [...this.tasks.values()];
     }
 
     async attach({ terminalId, remoteTerminal }: AttachTaskTerminalParams): Promise<void> {
@@ -113,11 +109,21 @@ export class GitpodTaskServerImpl implements GitpodTaskServer {
         terminalProcess.write(`${supervisorBin} terminal attach -ir ${remoteTerminal}\r\n`);
     }
 
-    setClient(client: GitpodTaskClient): void {
-        this.clients.add(client);
-    }
-    disposeClient(client: GitpodTaskClient): void {
-        this.clients.delete(client);
+    setClient(client: JsonRpcProxy<GitpodTaskClient>): void {
+        let closed = false;
+        this.deferredReady.promise.then(() => {
+            if (closed) {
+                return;
+            }
+            this.clients.add(client);
+            client.onDidChange({
+                updated: [...this.tasks.values()]
+            })
+        });
+        client.onDidCloseConnection(() => {
+            closed = true;
+            this.clients.delete(client);
+        });
     }
 
     dispose(): void {

